### PR TITLE
Avoid retrying on same instance

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
Fixes gh-747.

### CODE REVIEW

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 1]
```diff
@@ -448,6 +448,9 @@ You can set:
 - `spring.cloud.loadbalancer.retry.maxRetriesOnSameServiceInstance` - indicates how many times a request should be retried on the same `ServiceInstance` (counted separately for every selected instance)
 - `spring.cloud.loadbalancer.retry.maxRetriesOnNextServiceInstance` - indicates how many times a request should be retried a newly selected `ServiceInstance`
 - `spring.cloud.loadbalancer.retry.retryableStatusCodes` - the status codes on which to always retry a failed request.
+
+NOTE:: For load-balanced retries, by default, we instantiate `AvoidPreviousInstanceRoundRobinLoadBalancer` that will select a different instance from the one previously chosen if available. You can switch back to using the basic `RoundRobinLoadBalancer` by setting the value of `spring.cloud.loadbalancer.retry.avoidPreviousInstance.enabled` to `false`.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:452]: Change to "NOTE: For load-balanced retries, by default, we instantiate `AvoidPreviousInstanceRoundRobinLoadBalancer` to select a different instance from the one previously chosen, if available. You can switch back to using the basic `RoundRobinLoadBalancer` by setting the value of `spring.cloud.loadbalancer.retry.avoidPreviousInstance.enabled` to `false`."

**Review by Buzzardo — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/_configprops.adoc [hunk 2]
```diff
@@ -35,6 +35,7 @@
 |spring.cloud.loadbalancer.health-check.interval | 25s | Interval for rerunning the HealthCheck scheduler.
 |spring.cloud.loadbalancer.health-check.path |  | 
 |spring.cloud.loadbalancer.hint |  | Allows setting the value of <code>hint</code> that is passed on to the LoadBalancer request and can subsequently be used in {@link ReactiveLoadBalancer} implementations.
+|spring.cloud.loadbalancer.retry.avoidPreviousInstance.enabled | true | Instantiates AvoidPreviousInstanceRoundRobinLoadBalancer if Spring-Retry is in the classpath.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/_configprops.adoc:38]: Add back ticks around AvoidPreviousInstanceRoundRobinLoadBalancer

**Review by Buzzardo — CHANGES_REQUESTED** [review]

File: spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * A modified {@link RoundRobinLoadBalancer} implementation that avoids picking the same
+ * service instance while retrying requests.
+ *
+ * @author Olga Maciaszek-Sharma
+ * @since 3.0.0
+ * @see RetryLoadBalancerInterceptor
+ */
+public class AvoidPreviousInstanceRoundRobinLoadBalancer extends RoundRobinLoadBalancer {
```
Comment (spencergibb) [inline spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java:43]: I wonder if we can do this without tying ourselves to an algorithm?

Comment (spencergibb) [reply spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java:43]: It comes down to getting a `List<ServiceInstance>` eventually. What you've implemented is basically a `ServiceInstanceListSupplier.get(Request)` or maybe `.get(RequestContext)`.

Comment (spencergibb) [reply spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java:43]: I think that would work.

in `ServiceInstanceListSupplier`

```
  // for backwards compatibility
	default Flux<List<ServiceInstance>> get(Request<?> request) {
		return get();
	}
```

in a new `RetryAwareServiceInstanceListSupplier`
```
	public Flux<List<ServiceInstance>> get(Request req) {
		get().map(instanceList -> {
			//... this is the logic below from choose() and getInstanceResponse()
		});
	}
```
Then we can add `RetryAwareServiceInstanceListSupplier` to the builder and for the defaults where retry is used we automatically wrap the existing `ServiceInstanceListSupplier`

Comment (OlgaMaciaszek) [reply spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/AvoidPreviousInstanceRoundRobinLoadBalancer.java:43]: Ok. Will work on changing it that way.

File: docs/src/main/asciidoc/_configprops.adoc (hunk 2 above)
Comment (OlgaMaciaszek) [reply docs/src/main/asciidoc/_configprops.adoc]: This file is automatically generated.

**Review by spencergibb — APPROVED** [review]
besides property name case

File: spring-cloud-loadbalancer/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 4]
```diff
@@ -21,6 +21,12 @@
 			"name": "spring.cloud.loadbalancer.cache.enabled",
 			"description": "Enables Spring Cloud LoadBalancer caching mechanism.",
 			"type": "java.lang.Boolean"
+		},
+		{
+			"defaultValue": true,
+			"name": "spring.cloud.loadbalancer.retry.avoidPreviousInstance",
```
Comment (spencergibb) [inline spring-cloud-loadbalancer/src/main/resources/META-INF/additional-spring-configuration-metadata.json:27]: `avoid-previous-instance`

File: spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/annotation/LoadBalancerClientConfiguration.java [hunk 5]
```diff
@@ -129,4 +136,42 @@ public ServiceInstanceListSupplier healthCheckDiscoveryClientServiceInstanceList
...
+			super(ConfigurationPhase.REGISTER_BEAN);
+		}
+
+		@ConditionalOnProperty(value = "spring.cloud.loadbalancer.retry.enabled", havingValue = "true",
+				matchIfMissing = true)
+		static class LoadBalancerRetryEnabled {
+
+		}
+
+		@ConditionalOnProperty(value = "spring.cloud.loadbalancer.retry.avoidPreviousInstance", havingValue = "true",
```
Comment (spencergibb) [inline spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/annotation/LoadBalancerClientConfiguration.java:169]: `avoid-previous-instance`

File: docs/src/main/asciidoc/_configprops.adoc [hunk 6]
```diff
@@ -35,6 +35,7 @@
 |spring.cloud.loadbalancer.health-check.interval | 25s | Interval for rerunning the HealthCheck scheduler.
 |spring.cloud.loadbalancer.health-check.path |  | 
 |spring.cloud.loadbalancer.hint |  | Allows setting the value of <code>hint</code> that is passed on to the LoadBalancer request and can subsequently be used in {@link ReactiveLoadBalancer} implementations.
+|spring.cloud.loadbalancer.retry.avoid-previous-instance | true | Enables wrapping ServiceInstanceListSupplier beans with `RetryAwareServiceInstanceListSupplier` if Spring-Retry is in the classpath.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/_configprops.adoc:38]: Add back ticks around ServiceInstanceListSupplier

Also, all the values that are code should be wrapped in back ticks. That's true in this line, and I see code values in some of the other lines, too.

File: docs/src/main/asciidoc/spring-cloud-commons.adoc [hunk 7]
```diff
@@ -448,6 +448,9 @@ You can set:
 - `spring.cloud.loadbalancer.retry.maxRetriesOnSameServiceInstance` - indicates how many times a request should be retried on the same `ServiceInstance` (counted separately for every selected instance)
 - `spring.cloud.loadbalancer.retry.maxRetriesOnNextServiceInstance` - indicates how many times a request should be retried a newly selected `ServiceInstance`
 - `spring.cloud.loadbalancer.retry.retryableStatusCodes` - the status codes on which to always retry a failed request.
+
+NOTE: For load-balanced retries, by default, we wrap the `ServiceInstanceListSupplier` bean with `RetryAwareServiceInstanceListSupplier` to select a different instance from the one previously chosen, if available. You can disable this behaviour by setting the value of `spring.cloud.loadbalancer.retry.avoidPreviousInstance` to `false`.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-commons.adoc:452]: Take the u out of behaviour. (Spring uses American spellings.) My bad for not catching that earlier. I am too used to reading British English.

**Review by Buzzardo — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/_configprops.adoc (hunk 6 above)
Comment (OlgaMaciaszek) [reply docs/src/main/asciidoc/_configprops.adoc:38]: This file is automatically generated.

**Pull request merged**